### ISSUE TITLE
If future returns Pending, it must store waker

### ISSUE THREAD

**Issue by MikailBag** [rust-lang/rust-clippy#6314]
### What it does
If future returns Pending, it should use the provided waker to notify executor that it can be polled again.
If the waker is not stored, the future will be unable to notify it and therefore it will be never polled again. Such a future is probably incorrect.

### Categories (optional)

- Kind: correctness
### Drawbacks
False positive here is a future that sometimes diverges and sometimes terminates (because if a future always diverges that programmer should use `futures_util::future::pending()` instead). I think such features are rare, but don't have any estimations or experiments.
### Example

```rust
impl Future for MyFut {
    type Output = ...;
    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        match self.state() {}
   }
}
```
